foundryup/src/utils.rs

```diff
@@ -0,0 +1,282 @@
+//! common utilities used in foundryup
+use crate::{
+    errors::FoundryupError,
+    process::{get_process, with_default, Processor},
+};
+use eyre::WrapErr;
+use std::{
+    env, fs,
+    fs::File,
+    io,
+    io::Write,
+    path::{Path, PathBuf},
+};
+use url::Url;
+
+/// The version message for the current program, like
+/// `foundryup 0.1.0 (f01b232bc 2022-01-22T23:28:39.493201+00:00)`
+pub(crate) const VERSION_MESSAGE: &str = concat!(
+    env!("CARGO_PKG_VERSION"),
+    " (",
+    env!("VERGEN_GIT_SHA_SHORT"),
+    " ",
+    env!("VERGEN_BUILD_TIMESTAMP"),
+    ")"
+);
+
+/// New type for a process exit code
+pub struct ExitCode(pub i32);
+
+impl From<i32> for ExitCode {
+    fn from(code: i32) -> Self {
+        ExitCode(code)
+    }
+}
+
+/// Returns the home directory
+pub fn home_dir() -> Option<PathBuf> {
+    get_process().home_dir()
+}
+
+/// Returns the storage directory used by foundryup
+pub fn foundryup_dir() -> io::Result<PathBuf> {
+    Ok(foundry_home()?.join("foundryup"))
+}
+
+/// Returns the home directory used by foundry
+///
+/// It returns one of the following values, in this order of
+/// preference:
+///
+/// - The value of the `FOUNDRY_HOME` environment variable, if it is an absolute path.
+/// - The value of the current working directory joined with the value of the `FOUNDRY_HOME`
+///   environment variable, if `FOUNDRY_HOME` is a relative directory.
+/// - The `.foundry` directory in the user's home directory, as reported by the `home_dir` function.
+///
+/// # Errors
+///
+/// This function fails if it fails to retrieve the current directory,
+/// or if the home directory cannot be determined.
+pub fn foundry_home() -> io::Result<PathBuf> {
+    with_default(|env| foundry_home_from(&**env))
+}
+
+pub fn foundry_home_from(env: &dyn Processor) -> io::Result<PathBuf> {
+    let cwd = env.current_dir()?;
+    foundry_home_with_cwd_from(env, &cwd)
+}
+
+pub fn foundry_home_with_cwd_from(env: &dyn Processor, cwd: &Path) -> io::Result<PathBuf> {
+    if let Some(home) = env.var_os("FOUNDRY_HOME").filter(|v| !v.is_empty()) {
+        {
+            let home = PathBuf::from(home);
+            return if home.is_absolute() { Ok(home) } else { Ok(cwd.join(&home)) }
+        }
+    }
+    env.home_dir()
+        .map(|d| d.join(".foundry"))
+        .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "could not find foundry home dir"))
```

Comment:
Should we consider using https://docs.rs/directories/latest/directories/ instead to support the XDG base directory specification (https://github.com/foundry-rs/foundry/issues/1262) and put things in the appropriate folders on other platforms too?